components/backup/src/disk_snap.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+//! This module contains things about disk snapshot.
+
+use std::{
+    sync::{atomic::AtomicU64, Arc},
+    task::Poll,
+    time::Duration,
+};
+
+use futures::future;
+use futures_util::{
+    future::{BoxFuture, FutureExt},
+    sink::SinkExt,
+    stream::StreamExt,
+};
+use grpcio::{RpcStatus, WriteFlags};
+use kvproto::{
+    brpb::{
+        PrepareSnapshotBackupEventType as PEvnT, PrepareSnapshotBackupRequest as PReq,
+        PrepareSnapshotBackupRequestType as PReqT, PrepareSnapshotBackupResponse as PResp,
+    },
+    errorpb,
+    metapb::Region,
+};
+use raftstore::store::{
+    snapshot_backup::{
+        RejectIngestAndAdmin, SnapshotBrHandle, SnapshotBrWaitApplyRequest, UnimplementedHandle,
+    },
+    SnapshotBrWaitApplySyncer,
+};
+use tikv_util::warn;
+use tokio::sync::oneshot;
+use tokio_stream::Stream;
+
+type Result<T> = std::result::Result<T, Error>;
+
+enum Error {
+    NotInitialized,
+    LeaseExpired,
+    WaitApplyAborted,
+    RaftStore(raftstore::Error),
+}
+
+enum HandleErr {
+    AbortStream(RpcStatus),
+    SendErrResp(errorpb::Error),
+}
+
+pub struct ResultSink(grpcio::DuplexSink<PResp>);
+
+impl From<grpcio::DuplexSink<PResp>> for ResultSink {
+    fn from(value: grpcio::DuplexSink<PResp>) -> Self {
+        Self(value)
+    }
+}
+
+impl ResultSink {
+    async fn send(
+        mut self,
+        result: Result<PResp>,
+        error_extra_info: impl FnOnce(&mut PResp),
+    ) -> grpcio::Result<Self> {
+        match result {
+            // Note: should we batch here?
+            Ok(item) => self.0.send((item, WriteFlags::default())).await?,
+            Err(err) => match err.how_to_handle() {
+                HandleErr::AbortStream(status) => {
+                    self.0.fail(status.clone()).await?;
+                    return Err(grpcio::Error::RpcFinished(Some(status)));
+                }
+                HandleErr::SendErrResp(err) => {
+                    let mut resp = PResp::new();
+                    error_extra_info(&mut resp);
+                    resp.set_error(err);
+                    self.0.send((resp, WriteFlags::default())).await?;
+                }
+            },
+        }
+        Ok(self)
+    }
+}
+
+impl Error {
+    fn how_to_handle(self) -> HandleErr {
```

Comment:
```suggestion
    fn into_handle_error(self) -> HandleErr {
```

Or impl `core::convert::Into`.

It's better to use rust idioms.